fairness/wasserstein.py

```diff
@@ -0,0 +1,542 @@
+    ----------
+    ecdf : dict
+        Dictionary storing ECDF (Empirical Cumulative Distribution Function) objects for each sensitive modality.
+    eqf : dict
+        Dictionary storing EQF (Empirical Quantile Function) objects for each sensitive modality.
+
+    Methods
+    -------
+    _check_shape(y, x_ssa)
+        Check the shape and data types of input arrays y and x_ssa.
+    _check_mod(sens_val_calib, sens_val_test)
+        Check if modalities in test data are included in calibration data's modalities.
+    _check_epsilon(epsilon)
+        Check if epsilon (fairness parameter) is within the valid range [0, 1].
+    _get_mod(x_ssa)
+        Get unique modalities from the input sensitive attribute array.
+    _get_loc(x_ssa)
+        Get the indices of occurrences for each modality in the input sensitive attribute array.
+    _get_weights(x_ssa)
+        Calculate weights (probabilities) for each modality based on their occurrences.
+    _estimate_ecdf_eqf(y, x_ssa, sigma)
+        Estimate ECDF and EQF for each modality, incorporating random noise within [-sigma, sigma].
+
+    Notes
+    -----
+    This base class provides essential methods for Wasserstein distance-based fairness adjustment. It includes
+    methods for shape validation, modality checks, epsilon validation, modality extraction, localization of
+    modalities in the input data, weight calculation, and ECDF/EQF estimation with random noise.
+    """
+
+    def __init__(self):
+        self.ecdf = {}
+        self.eqf = {}
+
+    def _check_shape(y, x_ssa):
+        """
+        Check the shape and data types of input arrays y and x_ssa.
+
+        Parameters
+        ----------
+        y : array-like
+            Target values of the data.
+        x_ssa : array-like
+            Input samples representing the sensitive attribute.
+
+        Raises
+        ------
+        ValueError
+            If the input arrays have incorrect shapes or data types.
+        """
+        if not isinstance(x_ssa, np.ndarray):
+            raise ValueError('x_sa must be an array')
+
+        if not isinstance(y, np.ndarray):
+            raise ValueError('y must be an array')
+
+        if len(x_ssa) != len(y):
+            raise ValueError('x_sa and y should have the same length')
+
+        for el in y:
+            if not isinstance(el, float):
+                raise ValueError('y should contain only float numbers')
+
+    def _check_mod(sens_val_calib, sens_val_test):
+        """
+        Check if modalities in test data are included in calibration data's modalities.
+
+        Parameters
+        ----------
+        sens_val_calib : list
+            Modalities from the calibration data.
+        sens_val_test : list
+            Modalities from the test data.
+
+        Raises
+        ------
+        ValueError
+            If modalities in test data are not present in calibration data.
+        """
+        if not all(elem in sens_val_calib for elem in sens_val_test):
```

Comment:
simplify, and show what the actual error is: for example something like


```suggestion
value_differences = set(sens_val_test) - set(sens_val_calib)
if len(value_differences) != 0:
    print_something_with differences

```